File: webui/frontend/src/lib/agentKind.ts
/** Classify chat agents as API, CLI, or remote (REQ-49 / REQ-203). */

export type AgentKind = 'api' | 'cli' | 'remote'

const KINDS = new Set<AgentKind>(['api', 'cli', 'remote'])

/** Remote implementations — not a fifth user-facing kind (ADR-011). */
const REMOTE_IMPL_IDS = new Set([
  'herdr',
  'hermes',
  'omb',
  'rakazo',
  'openmausbot',
  'openmaus',
  'openmousbot',
  'rakoza',
  'open-swarm',
  'openswarm',
  'open_swarm',
])

export function isRemoteImplId(raw: string | null | undefined): boolean {
  const key = (raw ?? '').trim().toLowerCase()
  if (!key) return false
  if (key.startsWith('herdr:') || key.startsWith('remote:')) return true
  return REMOTE_IMPL_IDS.has(key)
}

export function classifyAgentKind(
  raw: string | null | undefined,
  explicit?: string | null,
): AgentKind {
  if (explicit && KINDS.has(explicit as AgentKind)) {
    return explicit as AgentKind
  }
  if (isRemoteImplId(explicit)) return 'remote'
  const text = (raw ?? '').trim().toLowerCase()
  if (text.startsWith('cli:')) return 'cli'
  if (
    text.startsWith('remote:') ||
    text.startsWith('placeholder:remote:') ||
    text.startsWith('herdr:') ||
    isRemoteImplId(text)
  ) {
    return 'remote'
  }
  return 'api'
}

/** True only for API-agent threads. CLI/remote sessions are owned outside swarm. */
export function canEditAgentMessages(
  raw: string | null | undefined,
  explicit?: string | null,
): boolean {
  return classifyAgentKind(raw, explicit) === 'api'
}

File: src/swarm/core/agent_kind.py
"""Classify chat agents as API, CLI, or remote (REQ-49).

API-agent threads are owned by Open Swarm and may be edited in place.
CLI and remote sessions are owned outside swarm — no edit.
"""

from __future__ import annotations

from typing import Literal

AgentKind = Literal["api", "cli", "remote"]

_VALID_KINDS = frozenset({"api", "cli", "remote"})


def classify_agent_kind(
    raw: str | None,
    *,
    explicit: str | None = None,
) -> AgentKind:
    """Return ``api``, ``cli``, or ``remote`` for an agent id / source.

    Explicit kind (from a roster or fixture) wins when it is one of the
    three values. Otherwise source-style prefixes are used:

    * ``cli:<name>`` → cli
    * ``remote:<name>`` / ``placeholder:remote:…`` → remote
    * everything else (including API blueprints such as ``cli_agent``) → api
    """
    if explicit in _VALID_KINDS:
        return explicit  # type: ignore[return-value]
    text = (raw or "").strip().lower()
    if text.startswith("cli:"):
        return "cli"
    if text.startswith("remote:") or text.startswith("placeholder:remote:"):
        return "remote"
    return "api"


def can_edit_agent_messages(
    raw: str | None,
    *,
    explicit: str | None = None,
) -> bool:
    """True only for API-agent threads (REQ-49)."""
    return classify_agent_kind(raw, explicit=explicit) == "api"
